spices/SPICE-0024-annotation-converters.adoc

```diff
@@ -0,0 +1,405 @@
+    }
+  }
+}
+----
+
+== Proposed Solution
+
+To better serve this use case, we propose adding a new type of converter based on property annotations alongside the existing class- and path-based converters.
+
+== Detailed design
+
+=== Pkl API
+
+A new `annotationConverters` property will be added to link:{uri-stdlib-BaseValueRenderer}[`BaseValueRenderer`].
+
+Annotation converters to apply before values are rendered.
+
+Annotation converters are checked for each property of rendered `Typed` values.
+If a property has an annotation for which a converter is defined, the converter function is called with the property name, the value of the annotation, and the property value.
+The converter function must return a `Pair` of the converted property name and value.
+
+Each of the property annotations of the enclosing class is checked in order, starting with the annotations (top to bottom) of the current class, then repeating for each parent class in the hierarchy.
+Multiple annotation converters can apply; converters for matching annotations are called in order, with the result of one being passed to the next.
+
+At most one converter is applied per annotation. The most specific class (according to class hierarchy) wins.
+
+Annotation converters can overlap with link:{uri-stdlib-BaseValueRenderer}#converters[`converters`] and apply first.
+
+[source,pkl]
+----
+abstract class BaseValueRenderer {
+  // ...
+
+  annotationConverters: Mapping<Class, (String, unknown, unknown) -> Pair<String, Any>>
+
+  // ...
+}
+----
+
+=== Usage Examples
+
+The same scenarios above become simple to implement and maintain with annotation converters:
+
+* *Transforming property names*
++
+[source,pkl]
+----
+class SnakeCase extends Annotation {
+  function transform(value: String): String =
+    value.replaceAllMapped(Regex("([a-z0-9])([A-Z])"), (match) ->
+      "\(match.groups[1].value)_\(match.groups[2].value.toLowerCase())"
+    )
+}
+
+class Bird {
+  @SnakeCase
+  birdName: String
+}
+
+output {
+  renderer {
+    annotationConverters {
+      [SnakeCase] = (propertyName, annotation, value) ->
+        Pair(annotation.transform(propertyName), value)
+    }
+  }
+}
+----
+* *Converting the same type in different ways*
++
+[source,pkl]
+----
+clientConfig: ClientConfig
+
+serverConfig: ServerConfig
+
+class ClientConfig {
+  @RenderDuration { unit = "s" }
+  timeoutS: Duration
+}
+
+class ServerConfig {
+  @RenderDuration { unit = "s" }
+  readTimeoutS: Duration
+  @RenderDuration { unit = "ms" }
+  readHeaderTimeoutMS: Duration
+  children: Listing<ServerConfig>
+}
+
+class RenderDuration {
+  unit: DurationUnit
+
+  function transform(value: Duration): Number = value.toUnit(unit).value
+}
+
+output {
+  renderer {
+    annotationConverters {
+      [RenderDuration] = (propertyName, annotation, value) ->
+        Pair(propertyName, annotation.transform(value))
+    }
+  }
+}
+----
+* *Validating properties in nested structures*
```

Comment:
While this is interesting to think about, I don't think that we should promote converters as a way to do validation.

1. our IDE tooling doesn't understand this type of logic
2. it's easy to "hold it wrong" (you must evaluate the `output.{text,bytes}` to get validation)
3. it's a typecheck that's not provided by the language's type system

By the way: you can define this type of check today with existing converters too! You don't need annotation converters:

```pkl
output {
  renderer {
    converters {
      [Typed] = (it) ->
        let (clazz = reflect.Class(it.getClass())
          throwIfAnyUnavailableProperties(it, clazz, softwareVersion)
    }
  }
}
```